input/sqs.go

```diff
@@ -216,25 +231,38 @@ func (s *SQS) Run(inch chan<- *baker.Data) error {
 		for _, url := range resp.QueueUrls {
 			wg.Add(1)
 			go func(url string) {
-				defer wg.Done()
-				s.pollQueue(url)
+				defer func() {
+					wg.Done()
+				}()
+				s.pollQueue(ctx, url)
 			}(*url)
 		}
 	}
 
-	go func() {
-		// pollQueue() never exits for now, but in case we decide that it should,
-		// the correct thing to do is to notify the gzipInput that we're done with
-		// file processing, so that the input will shut down and brings down the
-		// whole topology.
-		wg.Wait()
-		s.NoMoreFiles()
-	}()
-
-	<-s.Done
+	// The correct order of operation to cleanly stop the whole pipeline is the
+	// following:
+	//  - first we close the 'done' channel, this in turns cancel polling via
+	//    context cancellation
+	//  - we then wait for all the polling goroutines to end.
+	//  - After this point we are guaranteed to not receive any more files so
+	//    we notify the embedded S3input.
+	//  - now we ask S3Input to stop as soon as it has finished processing files
+	//  - finally Run exits after being signaled from S3Input that we can
+	<-s.done
+	cancel()
+	wg.Wait()
```

Comment:
what do you think about adding a test for this stop procedure?

Reply:
I agree about adding a test, maybe in another PR